cvmfs/sqlitemem.cc

```diff
@@ -0,0 +1,489 @@
+    reinterpret_cast<char *>(new_avail) + new_size);
+  if (succ_size >= 0) {
+    // Merge with succeeding block and remove the block from the list
+    AvailBlockCtl *succ_avail = reinterpret_cast<AvailBlockCtl *>(
+      reinterpret_cast<char *>(new_avail) + new_size);
+    AvailBlockCtl *next = succ_avail->GetNextPtr(arena_);
+    AvailBlockCtl *prev = succ_avail->GetPrevPtr(arena_);
+    prev->link_next = succ_avail->link_next;
+    next->link_prev = succ_avail->link_prev;
+    new_size += succ_size;
+    if (rover_ == succ_avail)
+      rover_ = head_avail_;
+  }
+
+  // Insert new free block at the end of the list
+  AvailBlockCtl *next = head_avail_;
+  AvailBlockCtl *prev = head_avail_->GetPrevPtr(arena_);
+  next->link_prev = new_avail->ConvertToLink(arena_);
+  prev->link_next = new_avail->ConvertToLink(arena_);
+  new_avail->link_next = head_avail_->ConvertToLink(arena_);
+  new_avail->link_prev = prev->ConvertToLink(arena_);
+
+  // Set new free block's boundaries
+  new_avail->size = new_size;
+  void *upper_tag =
+    reinterpret_cast<char *>(new_avail) + new_size - sizeof(AvailBlockTag);
+  new (upper_tag) AvailBlockTag(new_size);
+}
+
+
+/**
+ * Walks the list of available blocks starting from rover and allocates the
+ * first available spot that's large enough.  Puts the reserved block at the end
+ * of the available one and, if necessary, removes the available one from the
+ * list of free blocks.
+ */
+void *MemoryManager::MallocArena::Malloc(const uint32_t size) {
+  assert(size > 0);
+
+  // Control word first, block type tag last
+  int32_t total_size = sizeof(ReservedBlockCtl) + size + 1;
+  if (total_size < kMinBlockSize)
+    total_size = kMinBlockSize;
+
+  bool wrapped = false;
+  // Generally: p = LINK(q)
+  AvailBlockCtl *q = rover_;
+  AvailBlockCtl *p;
+  do {
+    p = q->GetNextPtr(arena_);
+    if (p->size >= total_size)
+      break;
+    if (p == head_avail_) {
+      if (wrapped)
+        return NULL;
+      wrapped = true;
+    }
+    q = p;
+  } while (true);
+
+  // As of here: we found a sufficiently large free block at p following q
+  no_reserved_++;
+
+  int32_t remaining_size = p->size - total_size;
+  // Avoid creation of very small blocks
+  if (remaining_size < kMinBlockSize) {
+    total_size += remaining_size;
+    remaining_size = 0;
+  }
+
+  // Update the list of available blocks
+  if (remaining_size == 0) {
+    // Remove free block p from the list of available blocks
+    q->link_next = p->link_next;
+    q->GetNextPtr(arena_)->link_prev = q->ConvertToLink(arena_);
+  } else {
+    p->size = remaining_size;
+    void *upper_tag =
+      reinterpret_cast<char *>(p) + remaining_size - sizeof(AvailBlockTag);
+    new (upper_tag) AvailBlockTag(remaining_size);
+  }
+  rover_ = q->GetNextPtr(arena_);
+
+  // Place the new allocation, which also sets the block type tag at the end
+  char *new_block = reinterpret_cast<char *>(p) + remaining_size;
+  new (new_block) ReservedBlockCtl(total_size);
```

Comment:
After `FindFreeBlock()` the rest of the method would then be `ReserveBlock()`. I claim that way it would be a lot easier to understand the subparts, given the amount of pointer arithmetic in that method.